### ISSUE TITLE
About training and test

### ISSUE THREAD

**Issue by TomHJ** [nithi89/unet_darknet#2]
Hi, Thanks for your open project. I am checking if semantic segmentation is available under darknet. i cannot find it until i saw this project. thanks very much!

- How about your test speed? what it is the FPS and your picture's resolution, hardware configuration?

- Is it OK for many number of instances such as 15 classes of instance need to be tested?

- Do you have some material or article to introduce how to train special dataset?

**Comment by bonsai89** [comment]
Hi, I just have one output class, the prediction speed is 18ms on average on GeForce Titan X 12Gb. You can add multiple output classes, you will have to modify the cfg file accordingly.

I don't have an article, as the test setup and train is the same as all of the examples in darknet.
I added some additional notes to Readme.

**Comment by TomHJ** [comment]
Thanks for your words! I am studying how to train custom data, some questions about label mark as below:
How about the input image format? Are .bmp or .jpg available? 
Does images need to be resized into 224x224x3 before label mark? 
Is the label mark software "Labelme" available? Because output of Labelme is .png file with int32, need to convert it from int32 to int8? 

Thanks

**Comment by bonsai89** [comment]
The images are loaded using opencv in darknet, so all formats supported by opencv should be workable.

The images need to be resized to 224x224x3 before labeling or you need to change the input size to the network in the unet.cfg file to match your input image.

I use matlab ROI function to generate the binary mask labels.

**Comment by TomHJ** [comment]
Thanks for your words above! Some question about the training: Do you train from scratch or use transfer learning(using  pre-trained weights)? Can you share it if possible? To me, it is very hard to train from scratch by my estimation.How many iterations have you trained? What is the value of loss?

**Comment by TomHJ** [comment]
One question about label image: It seems some data set of semantic segmentation whose label images have different channels(cityscapes dataset: 8-bit gray scale or 16-bit gray scale or 8-bit/color RGBA), and in your project whose label images is 8-bit color RGB file which is three channel. I do not know the difference.

**Comment by bonsai89** [comment]
In my project the label is single channel gray scale image. As to answer your previous question I train my network from scratch and it took about 800 iterations to get reasonable performance.

**Comment by TomHJ** [comment]
Thanks for your reply, it is useful for me to understand the label images. 
Sorry, i have another question here which made me stuck at training.  Below is console output.Appreciate any comment!

Problem:  
1.predication was always NAN, loss was always NAN ( from iteration 1 ~ iteration 1500,batchsize was 1,cpu mode)

2. It seems image and its label loaded normally, i checked some t value were 1.000000

3.It was strange when changed training samples from 684 to 16,  t value were abnormal which seems should be pred value as below showing. based on my understanding: t value should be label value(1.000000 or 0.0000000)
........................
t 94.000000,p nan 
err nan, del nan 
t 94.000000,p nan 
err nan, del nan 
t 95.000000,p nan 
err nan, del nan 
t 94.000000,p nan 
err nan, del nan 
t 95.000000,p nan 
err nan, del nan 
my cost is nan
.....................

4. I tried to debug the program, it seemed that 1st layer's output was nan, also i found 1st layer's weights and bias was nan, i did not know why. And the debug was on going.

Below is training output from console(training samples: 684)
``````````````````````````````````````````````````````````````````````````````````````````````````````````
tom@tom-ML:~/segmentation/unet_darknet$ ./darknet unet_segmenter train experiment_0924/parameter/unet.data experiment_0924/parameter/unet.cfg experiment_0924/backup/unet.backup 
subdivisions: Using default '1'
policy: Using default 'constant'
layer     filters    size              input                output
    0 conv     32  3 x 3 / 1   224 x 224 x   3   ->   224 x 224 x  32  0.087 BFLOPs
    1 conv     32  3 x 3 / 1   224 x 224 x  32   ->   224 x 224 x  32  0.925 BFLOPs
    2 max          2 x 2 / 2   224 x 224 x  32   ->   112 x 112 x  32
    3 conv     64  3 x 3 / 1   112 x 112 x  32   ->   112 x 112 x  64  0.462 BFLOPs
    4 conv     64  3 x 3 / 1   112 x 112 x  64   ->   112 x 112 x  64  0.925 BFLOPs
    5 max          2 x 2 / 2   112 x 112 x  64   ->    56 x  56 x  64
    6 conv    128  3 x 3 / 1    56 x  56 x  64   ->    56 x  56 x 128  0.462 BFLOPs
    7 conv    128  3 x 3 / 1    56 x  56 x 128   ->    56 x  56 x 128  0.925 BFLOPs
    8 max          2 x 2 / 2    56 x  56 x 128   ->    28 x  28 x 128
    9 conv    256  3 x 3 / 1    28 x  28 x 128   ->    28 x  28 x 256  0.462 BFLOPs
   10 conv    256  3 x 3 / 1    28 x  28 x 256   ->    28 x  28 x 256  0.925 BFLOPs
   11 max          2 x 2 / 2    28 x  28 x 256   ->    14 x  14 x 256
   12 conv    512  3 x 3 / 1    14 x  14 x 256   ->    14 x  14 x 512  0.462 BFLOPs
   13 conv    512  3 x 3 / 1    14 x  14 x 512   ->    14 x  14 x 512  0.925 BFLOPs
   14 max          2 x 2 / 2    14 x  14 x 512   ->     7 x   7 x 512
   15 conv   1024  3 x 3 / 1     7 x   7 x 512   ->     7 x   7 x1024  0.462 BFLOPs
   16 conv   1024  3 x 3 / 1     7 x   7 x1024   ->     7 x   7 x1024  0.925 BFLOPs
   17 upsample            2x     7 x   7 x1024   ->    14 x  14 x1024
   18 route  17 13
   19 conv    512  3 x 3 / 1    14 x  14 x1536   ->    14 x  14 x 512  2.775 BFLOPs
   20 conv    512  3 x 3 / 1    14 x  14 x 512   ->    14 x  14 x 512  0.925 BFLOPs
   21 upsample            2x    14 x  14 x 512   ->    28 x  28 x 512
   22 route  21 10
   23 conv    256  3 x 3 / 1    28 x  28 x 768   ->    28 x  28 x 256  2.775 BFLOPs
   24 conv    256  3 x 3 / 1    28 x  28 x 256   ->    28 x  28 x 256  0.925 BFLOPs
   25 upsample            2x    28 x  28 x 256   ->    56 x  56 x 256
   26 route  25 7
   27 conv    128  3 x 3 / 1    56 x  56 x 384   ->    56 x  56 x 128  2.775 BFLOPs
   28 conv    128  3 x 3 / 1    56 x  56 x 128   ->    56 x  56 x 128  0.925 BFLOPs
   29 upsample            2x    56 x  56 x 128   ->   112 x 112 x 128
   30 route  29 4
   31 conv     64  3 x 3 / 1   112 x 112 x 192   ->   112 x 112 x  64  2.775 BFLOPs
   32 conv     64  3 x 3 / 1   112 x 112 x  64   ->   112 x 112 x  64  0.925 BFLOPs
   33 upsample            2x   112 x 112 x  64   ->   224 x 224 x  64
   34 route  33 1
   35 conv     32  3 x 3 / 1   224 x 224 x  96   ->   224 x 224 x  32  2.775 BFLOPs
   36 conv     32  3 x 3 / 1   224 x 224 x  32   ->   224 x 224 x  32  0.925 BFLOPs
   37 conv      1  1 x 1 / 1   224 x 224 x  32   ->   224 x 224 x   1  0.003 BFLOPs
   38 logistic x entropy                             50176
Loading weights from experiment_0924/backup/unet.backup...Done!
unet
1
-1 pointer 12
Learning Rate: 1e-07, Momentum: 0.9, Decay: 0.0005
/home/tom/segmentation/unet_darknet/experiment_0924/data/train/2520_4643_1711_s4_NG_7.png
came here path! 618 ,/home/tom/segmentation/unet_darknet/experiment_0924/data/train/2520_4643_1711_s4_NG_7.png
came here label! 618 , /home/tom/segmentation/unet_darknet/experiment_0924/data/labels/2520_4643_1711_s4_NG_7.png
Loaded: 0.004034 seconds
t 0.000000,p nan 
err nan, del nan 
t 0.000000,p nan 
err nan, del nan 
t 0.000000,p nan 
err nan, del nan 
t 0.000000,p nan 
err nan, del nan 
t 0.000000,p nan 
err nan, del nan 
t 0.000000,p nan 
err nan, del nan 
t 0.000000,p nan 
err nan, del nan 
t 0.000000,p nan 
err nan, del nan 
t 0.000000,p nan 
err nan, del nan 
t 0.000000,p nan 
err nan, del nan 
t 0.000000,p nan 
err nan, del nan 
t 0.000000,p nan 
...........................
...........................
............................
t 0.000000,p nan 
err nan, del nan 
my cost is nan 
n bathc is nan
1502, 2.196: nan, nan avg, 0.000000 rate, 19.174776 seconds, 1502 images
/home/tom/segmentation/unet_darknet/experiment_0924/data/train/2520_3415_1548_s4_NG_29.png
came here path! 424 ,/home/tom/segmentation/unet_darknet/experiment_0924/data/train/2520_3415_1548_s4_NG_29.png
came here label! 424 , /home/tom/segmentation/unet_darknet/experiment_0924/data/labels/2520_3415_1548_s4_NG_29.png
Loaded: 0.003936 seconds
.................................
..................................

**Comment by zouwen198317** [comment]
i want to  multiple output classes,
How  modify the  cfg file  ,tks

**Comment by bonsai89** [comment]
Please look at the other example models in darknet original examples and cfg folder. 
They have examples that cover different scenarios. 

As for the nan problem, I faced similar issue and I solved it by following through the code line by line and identifying the function where it is becoming nan, in my case it was division by zero (float values getting rounded to zero). It is quite hairy but you have to do it the hard way!

**Comment by tuuzhang** [comment]
@TomHJ @nithi89 Hi, guys, I want recurring the experimental results，but failed! I used the original picture (5 png pictures) to train the model, but the  p value was still nan and err was also still nan. How to debug it?
